core/server/common/src/main/java/alluxio/master/journal/JournalUtils.java

```diff
@@ -214,5 +220,123 @@ public static void sinkFlush(Supplier<Set<JournalSink>> journalSinks) {
     }
   }
 
+  /**
+   * Merge inode entry with subsequent update inode and update inode file entries.
+   *
+   * @param entries list of journal entries
+   * @return a list of compacted journal entries
+   */
+  public static List<alluxio.proto.journal.Journal.JournalEntry> mergeCreateComplete(
+      List<alluxio.proto.journal.Journal.JournalEntry> entries) {
+    List<alluxio.proto.journal.Journal.JournalEntry> newEntries = new ArrayList<>();
+    Map<Long, File.InodeFileEntry.Builder> fileEntryMap = new HashMap<>();
+    for (alluxio.proto.journal.Journal.JournalEntry oldEntry : entries) {
+      if (oldEntry.hasInodeFile()) {
+        fileEntryMap.put(oldEntry.getInodeFile().getId(),
+            File.InodeFileEntry.newBuilder(oldEntry.getInodeFile()));
+      } else if (oldEntry.hasUpdateInode()) {
+        File.UpdateInodeEntry entry = oldEntry.getUpdateInode();
```

Comment:
Would it be possible to reuse the update logic from `MutableInode::updateFromEntry`? It can be challenging to keep multiple places in sync and might result in missing updates when later on someone add a new field to the entry.

Reply:
yeah i thought about that as well, but it is actually a bit tricky because of the subtle difference between the inode class and the inode proto.  They do not share anything in the inheritance tree either, so that makes it difficult to factor out this part of the logic.  any suggestions?

Reply:
Would it be easier to convert entry to `MutableInodeFile` first? Something like

```
    Map<Long, Pair<Integer, MutableInodeFile>> fileEntryMap = new HashMap<>();
    for (alluxio.proto.journal.Journal.JournalEntry oldEntry : entries) {
      if (oldEntry.hasInodeFile()) {
        ...
        fileEntryMap.put(oldEntry.getInodeFile().getId(),
            new Pair<>(newEntries.size() - 1,
                MutableInodeFile.fromJournalEntry(oldEntry.getInodeFile())));
      } else if (oldEntry.hasUpdateInode()) {
        ...
        MutableInodeFile inode = fileEntryMap.get(entry.getId()).getSecond();
        inode.updateFromEntry(entry);
      } else if (oldEntry.hasUpdateInodeFile()) {
        ...
        MutableInodeFile inode = fileEntryMap.get(entry.getId()).getSecond();
        inode.updateFromEntry(entry);
```